### DISCUSSION TITLE
grammar to distinguish between identifier and implicit assignment

### DISCUSSION THREAD

**Discussion by ridilculous** [tree-sitter/tree-sitter#4700] [Q&A] [answered]
I'm stuck trying to create a grammar which can distinguish between a assignment expression with a explicit assignment operator, like

    T1=1

and an implicit assignment like

    T1

where the variable names for the explicit assignment can have trailing numbers and the ones for implicit assignment cannot, i.e. the former should assign 1 to T1 and the latter 1 to T.

I came up with the following simple grammar:

```javascript
module.exports = grammar({
	name: 'test',
	rules: {
		root: $ => repeat($.a),

		a: $ => choice($.i, $.e),

		i: $ => seq($.s, $.n),
		e: $ => seq($.sn, '=', $.n),

		s: $ => /[A-Z]+/,
		sn: $ => /[A-Z]+[0-9]*/,

		n: $ => /[0-9]+/
	}
});
```

This works for these test cases

    T 1
    T1=1
    T1 = 1

but doesn't for

    T1

or

    T=1

This is the test file i'm working with:

    ====================================
    works
    ====================================
    T 1
    T1=1
    T1 = 1
    ---
    (root
        (a (i (s)  (n)))
        (a (e (sn) (n)))
        (a (e (sn) (n))))
    ====================================
    e1) doesn't work
    ====================================
    T1
    ---
    (root (a (i (s)  (n))))
    ====================================
    e2) doesn't work either
    ====================================
    T=1
    ---
    (root (a (e (sn) (n))))
    ====================================
    e1) results in
    (root (ERROR (sn)))
    and
    e2) in
    (root (a (i (s) (ERROR) (n))))
    ====================================

I tried to tinker with precedence, f.e. give $.i a higher number, but nothing helped so far.
I don't think the grammar is ambiguous but maybe i just don't see it clearly?

Is it possible to express such a grammar with tree-sitter?

_I also postet this [on SO](https://stackoverflow.com/questions/70855891/tree-sitter-grammar-to-distinguish-between-identifier-and-implicit-assign)._

**Comment by ahelwer** [comment]
Have you read the conflicting token rules? Lexical precedence is different from parser preference.

**Comment by ridilculous** [comment]
I've read [them](https://tree-sitter.github.io/tree-sitter/creating-parsers#conflicting-tokens) multiple times but to be honest i'm not sure if i understand the implications. I also can't understand what you are hinting at. Is it that it's on the lexing side?

I now also tried to use the conflicts feature (also as hinted [here](https://github.com/tree-sitter/tree-sitter/issues/233)) but wasn't successful so far or just couldn't derive how to apply them and so more or less blindly tried things like

```javascript
conflicts: $ => [
	[$.s],
	[$.n],
	[$.sn],
	[$.s, $.n],
	[$.sn, $.n],
],
```
together with trying to increase `prec.dynamic` of `s` but it didn't click yet.

**Comment by ahelwer** [comment]
The conflicts field triggers GLR parsing (affected by prec.dynamic) and the prec function affects parser precedence. Have you tried using prec inside the token function to affect lexical precedence? Also, running tree sitter parse -d is often good to get an idea of what your parser is doing.

**Comment by ridilculous** [comment]
I hope i got this right. I removed all conflicts entries and tried different _token(prec(_ combinations (i wasn't aware of this possibility before). Now

```javascript
s: $ => token(prec(0,/[A-Z]+/)),
sn: $ => token(prec(1,/[A-Z]+[0-9]*/)),
```
errors for `T1`

```javascript
s: $ => token(prec(2,/[A-Z]+/)),
sn: $ => token(prec(1,/[A-Z]+[0-9]*/)),
```
errors for `T=1`

and
```javascript
s: $ => token(prec(1,/[A-Z]+/)),
sn: $ => token(prec(1,/[A-Z]+[0-9]*/)),
```
errors for both.

**Comment by alemuller** [comment, marked as answer]
You need to break the token to be able to use GLR. 

```js
module.exports = grammar({
  name: 'test',

  conflicts: $ => [
    [$.s, $.sn],
  ],

  rules: {
    root: $ => repeat($.a),

    a: $ => choice($.i, $.e),

    i: $ => seq($.s, $.n),
    e: $ => seq($.sn, '=', $.n),

    s: $ => $._IDENT,
    sn: $ => seq($._IDENT, optional(/[0-9]+/)),

    n: $ => /[0-9]+/,

    _IDENT: $ => /[A-Z]+/,

  }
});
```

@ahelwer This is the similar to #1615. Both issues could be easily solved if there was a way to annotate tokens conflicts to be resolved with GLR.

**Comment by ridilculous** [comment]
Great! This solves almost all problems.
Now i remember why i might have put the two together in `sn`;
`T 1 = 1`
is not supposed to be valid.
I tried to play around with the sn sequence like
```javascript
sn: $ => seq($._IDENT, optional(/[0-9]+/)),
//sn: $ => seq($._IDENT, token.immediate(optional(/[0-9]+/))),
//sn: $ => seq($._IDENT, optional(token.immediate(/[0-9]+/))),
```
but all attempts broke at least one other test.

I never felt comfortable using `token.immediate` because it always felt like i'm not getting its point.
Would maybe clearing `extras` and handling whitespaces explicitly help?

Here's the full test set:
```
====================================
works and is expected to work
====================================
T1
T=1
T1=1
T 1
T1 = 1
---
====================================
works but shouldn't
====================================
T 1=1
T 1 = 1
---
```

_Sidenode:_ I didn't notice it before but was always under the impression that
`sn: $ => seq($._IDENT, optional(/[0-9]+/))`
and
`sn: $ => seq($._IDENT, /[0-9]*/)`
would be equal but there's clearly a difference.

**Comment by alemuller** [comment]
> Would maybe clearing extras and handling whitespaces explicitly help?

It would. Whether it makes sense, depends on the language.

> I never felt comfortable using token.immediate because it always felt like i'm not getting its point.

Immediate tokens doesn't match when preceded by extras.

>  I tried to play around with the sn sequence like ...

Having `/[0-9]+/` and `token.immediate(/[0-9]+/)` returns to the original problem (the immediate token will always match when it is immediate).

People often use the external scanner to deal with those cases. It is not impossible to deal with them in the grammar, but it more complex (need to have a solid understanding on how the precedence works).

**Comment by ahelwer** [comment]
@ridilculous tree-sitter's design philosophy is to be permissive in what syntax it accepts, which is often useful for its primary use cases of highlighting and language tooling generally. It is much more difficult to create a tree-sitter grammar that only accepts your language's syntax rather than a superset of that syntax.
